## Imports

In the code cell below import the following 
- pandas as __pd__, 
- numpy as __np__, and
- matplotlib.pyplot as __plt__.
- seaborn as __sns__.

In [55]:
# import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


## Activity 1

Write code to perform following:
1. Mount your drive.
2. Read in the dataset saved in **Cohort_X/Data/Brazilian-fire-dataset.csv** and save it to a variable called `df_ffire`.
3. Display the top 10 rows of the data frame.
4. Generate summary statistics for all columns.

In [56]:
# mounting drive
from google.colab import drive
drive.mount('/drive')

Drive already mounted at /drive; to attempt to forcibly remount, call drive.mount("/drive", force_remount=True).


In [57]:
# read in data
df_ffire = pd.read_csv("/drive/My Drive/Cohort_7/Data/Brazilian-fire-dataset.csv")

In [58]:
# create a copy of the dataset to manipulate
df_ffire_copy = df_ffire.copy()

In [59]:
# display top 10 rows of the data frame
df_ffire_copy.head(10)

Year State    Month  Number of Fires Date Reported
0  1998  Acre  January              0.0     1/01/1998
1  1999  Acre  January              0.0     1/01/1999
2  2000  Acre  January              0.0     1/01/2000
3  2001  Acre  January              0.0     1/01/2001
4  2002  Acre  January              0.0     1/01/2002
5  2003  Acre  January             10.0     1/01/2003
6  2004  Acre  January              0.0     1/01/2004
7  2005  Acre  January             12.0     1/01/2005
8  2006  Acre  January              4.0     1/01/2006
9  2007  Acre  January              0.0     1/01/2007

In [60]:
# generate summary statistics for all columns
df_ffire_copy.describe(include= 'all')

Year    State    Month  Number of Fires Date Reported
count   6454.000000     6454     6454      6454.000000          6454
unique          NaN       27       12              NaN            20
top             NaN  Alagoas  January              NaN     1/01/2013
freq            NaN      240      541              NaN           324
mean    2007.461729      NaN      NaN       108.293163           NaN
std        5.746654      NaN      NaN       190.812242           NaN
min     1998.000000      NaN      NaN         0.000000           NaN
25%     2002.000000      NaN      NaN         3.000000           NaN
50%     2007.000000      NaN      NaN        24.000000           NaN
75%     2012.000000      NaN      NaN       113.000000           NaN
max     2017.000000      NaN      NaN       998.000000           NaN

In [61]:
# finding the amount of states that fire incidents have been reported
df_ffire_copy['State'].value_counts()

Alagoas                240
Amapá                  239
Acre                   239
Rio Grande do Norte    239
Pernambuco             239
Rio Grande do Sul      239
Paraíba                239
Roraima                239
Mato Grosso do Sul     239
Amazonas               239
Maranhão               239
Mato Grosso            239
Espírito Santo         239
São Paulo              239
Rio de January         239
Bahia                  239
Piauí                  239
Ceará                  239
Pará                   239
Santa Catarina         239
Goiás                  239
Minas Gerais           239
Sergipe                239
Distrito Federal       239
Tocantins              239
Rondônia               239
Paraná                 239
Name: State, dtype: int64

Based on the information collected so far, can you answer the following:

1.  What is the maximum number of fires reported on any given day?
2. In the dataset what is the total number of states for which the fire incidents have been reported?

1. The maximum number of fires reported on any given day is apparently 998 
1. In the dataset the total number of states for which fire incidents have been reported is 27


## Activity 2


1. Change the `Date Reported` column to a datetime object. 
2. Filter the `df_ffire` data frame to only keep the rows that are in the Acre, Bahia or Mato Grosso states. Save this filtered data frame as `df_ffire_filt`. 
3. Create a seaborn line plot of `df_ffire_filt` with the `Date Reported` on the x-axis, the `Number of Fires` on the y-axis and the `State` as the hue. 

In [62]:
# convert argument to datetime
df_ffire_copy['Date Reported'] = pd.to_datetime(df_ffire_copy['Date Reported'])

In [63]:
# checking to make sure I successfully changed the argument type
df_ffire_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6454 entries, 0 to 6453
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Year             6454 non-null   int64         
 1   State            6454 non-null   object        
 2   Month            6454 non-null   object        
 3   Number of Fires  6454 non-null   float64       
 4   Date Reported    6454 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(1), object(2)
memory usage: 252.2+ KB


In [64]:
# get the state names, so I can drop all but Acre, Bahia and Mato Grosso
df_ffire_copy['State'].value_counts()

Alagoas                240
Amapá                  239
Acre                   239
Rio Grande do Norte    239
Pernambuco             239
Rio Grande do Sul      239
Paraíba                239
Roraima                239
Mato Grosso do Sul     239
Amazonas               239
Maranhão               239
Mato Grosso            239
Espírito Santo         239
São Paulo              239
Rio de January         239
Bahia                  239
Piauí                  239
Ceará                  239
Pará                   239
Santa Catarina         239
Goiás                  239
Minas Gerais           239
Sergipe                239
Distrito Federal       239
Tocantins              239
Rondônia               239
Paraná                 239
Name: State, dtype: int64

In [68]:
# filtering the dataset to only keep the rows that are in the Acre, Bahia or Mato Grosso states
df_ffire_filt = df_ffire_copy[df_ffire_copy['State'].isin(['Acre', 'Bahia', 'Mato Grosso'])]

In [69]:
df_ffire_filt

Year        State     Month  Number of Fires Date Reported
0     1998         Acre   January            0.000    1998-01-01
1     1999         Acre   January            0.000    1999-01-01
2     2000         Acre   January            0.000    2000-01-01
3     2001         Acre   January            0.000    2001-01-01
4     2002         Acre   January            0.000    2002-01-01
...    ...          ...       ...              ...           ...
2625  2012  Mato Grosso  December          492.000    2012-01-01
2626  2013  Mato Grosso  December          447.000    2013-01-01
2627  2014  Mato Grosso  December          466.000    2014-01-01
2628  2015  Mato Grosso  December            1.777    2015-01-01
2629  2016  Mato Grosso  December          425.000    2016-01-01

[717 rows x 5 columns]

In [66]:
sns.lineplot(data=df_ffire_filt, x='Date Reported', y='Number of Fires', hue='State')

ValueError: ignored

## Activity 3
1. Find which state has recorded the maximum number of fires on any given day (not the total sum), and in which month did it occur?

HINT: 
- Create a boolean filter based on condition on column __Number of Fires__, and 
- Using this filter perform an indexing operation on the DataFrame to retrieve the corresponding records

In [ ]:
df_ffire_filt['State'].max()

## Activity 4

1. Calculate the total number of fires for the entire duration (i.e. for 1998 to 2018) for every state. Save the result to a data frame called `df_group_states`.
2. Print the Summary Statistics for `df_group_states` - and what is the minimum and maximum number of total fires records?
4. Which state has recorded the minimum number of total fires, and which state has recorded the maximum number of total fires?

__HINT:__
- Use the `groupby` operation on the State column. 
- Pandas has methods `idxmax()` and `idxmin()` which return the indices corresponding to the maximum and minimum values respectively, in a Pandas Series (or a Column in DataFrame).

In [ ]:
df_group_states = df_ffire.groupby('State')['Number of Fires'].sum()
df_group_states

In [ ]:
df_group_states.describe()

## Activity 5

1. Reset the index of the `df_group_states` DataFrame so that States are not considered as index, and are rather considered as Data Column (Hint: use method __reset_index()__ for this operation
2. Arrange the DataFrame in descending order of "Total Number of Fires".
3. Create a seaborn barplot with the State on the x-axis and the Number of Fires on the y-axis. Rotate the State labels on the x-axis so they are easier to read. 
    


In [ ]:
plt.figure(figsize = (10,7))
plt.xticks(rotation = 90)
sns.barplot(x = df_group_states['States'], y = df_group_states['Number of Fires'])